x-pack/plugins/maps/public/layers/sources/xyz_tms_source/xyz_tms_editor.tsx

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { Fragment, Component, ChangeEvent } from 'react';
+import _ from 'lodash';
+import { EuiFormRow, EuiFieldText } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+
+interface AttributionUpdate {
+  attributionText?: string;
+  attributionUrl?: string;
+}
+export interface XYZTMSSourceConfig extends AttributionUpdate {
+  urlTemplate: string;
+}
+
+export interface Props {
+  onSourceConfigChange: (sourceConfig: XYZTMSSourceConfig) => void;
+}
+
+interface State {
+  tmsInput: string;
+  tmsCanPreview: boolean;
+  attributionText?: string;
+  attributionUrl?: string;
+}
+
+export class XYZTMSEditor extends Component<Props, State> {
+  state = {
+    tmsInput: '',
+    tmsCanPreview: false,
+    attributionText: '',
+    attributionUrl: '',
+  };
+
+  _sourceConfigChange = _.debounce((updatedSourceConfig: XYZTMSSourceConfig) => {
+    if (this.state.tmsCanPreview) {
+      this.props.onSourceConfigChange(updatedSourceConfig);
+    }
+  }, 2000);
+
+  _handleTMSInputChange(e: ChangeEvent<HTMLInputElement>) {
+    const url = e.target.value;
+
+    const canPreview =
```

Comment:
Not sure if this is in scope for PR, but why is `tmsCanPreview` even part of state? `_sourceConfigChange` should just check that `tmsInput` is valid and avoid all this extra fluff

Reply:
handle in different PR (see below)